### DISCUSSION TITLE
configuration item `server_names` does not seem to work as intended (for me)

### DISCUSSION THREAD

**Discussion by jeroenubbink** [DNSCrypt/dnscrypt-proxy#2518] [Q&A] [answered]
I am running dnscrypt-proxy 2.1.4 on OpenBSD 7.3 and installed the binary package from openbsd's package repository.

I would like to filter out as many ads and trackers as possible, so i have setup my `server_names` configuration item to include several ad-blocking dns servers from the list in the commented out section just above the `server_names` in the configuration file. However from what i can tell it only ever picks the first one in the list. As soon as that become unavailable for whatever reason it refuses to pick another out of the list.

I am wondering if somebody else is experiencing the same problem and if somebody could maybe provide me with details on how i can debug this in order to try to get it fixed. Having one DNS server proves to be very unreliable.

my setting right now (added 2 non ad-blocking servers to confirm the behavior)

`server_names = ['brahma-world-ipv6', 'dnscrypt-de-blahdns-ipv6', 'cs-nl2', 'cs-de']`

The second entry used to be the first entry and it is down right now, if i put it as first entry again no resolving is possible.

**Comment by lifenjoiner** [comment, marked as answer]
The logic is: adopting servers follows `lb_strategy`, by delay ranking on the fly.

And I can't reproduce your result:
![2518](https://github.com/DNSCrypt/dnscrypt-proxy/assets/19585474/954de75e-f6ac-468f-8c6a-403d7b442397)

**Reply by jeroenubbink** [reply]
Ok so i went ahead and changed the following in configuration items because i had no logging of any kind:

```
server_names = ['dnscrypt-de-blahdns-ipv6', 'brahma-world-ipv6']
log_file = '/var/log/dnscrypt-proxy.log'
```

Restarted dnscrypt-proxy and now i can see in the log file that it does exactly the same for you as it does for me. Now i'm starting to wonder if not maybe all of the defined servers i had in there might have been down by chance. Everything looks good..

```
% cat /var/log/dnscrypt-proxy.log 
[2023-11-03 08:28:25] [NOTICE] dnscrypt-proxy 2.1.4
[2023-11-03 08:28:25] [NOTICE] Network connectivity detected
[2023-11-03 08:28:25] [NOTICE] Dropping privileges
[2023-11-03 08:28:25] [NOTICE] Network connectivity detected
[2023-11-03 08:28:25] [NOTICE] Now listening to 127.0.0.1:40 [UDP]
[2023-11-03 08:28:25] [NOTICE] Now listening to 127.0.0.1:40 [TCP]
[2023-11-03 08:28:25] [NOTICE] Now listening to [::1]:40 [UDP]
[2023-11-03 08:28:25] [NOTICE] Now listening to [::1]:40 [TCP]
[2023-11-03 08:28:25] [NOTICE] Source [public-resolvers] loaded
[2023-11-03 08:28:25] [NOTICE] Firefox workaround initialized
[2023-11-03 08:28:28] [NOTICE] [dnscrypt-de-blahdns-ipv6] TIMEOUT
[2023-11-03 08:28:28] [NOTICE] [brahma-world-ipv6] OK (DoH) - rtt: 11ms
[2023-11-03 08:28:28] [NOTICE] Server with the lowest initial latency: brahma-world-ipv6 (rtt: 11ms)
[2023-11-03 08:28:28] [NOTICE] dnscrypt-proxy is ready - live servers: 1
```

Additionally i found out that `cs-nl2` and `cs-de` are actually no longer on the public list so i guess i should have checked that first. It would be great if there was feedback about that in the log file because i tried again with the old configuration and it basically gave the same output, then i confirmed that these servers actually no longer exist.

Seems it was mostly due to most of these servers no longer existing anymore. They seem to appear and disappear quite rapidly (within a (few) year(s)).

The solution for me was to activate logging and just check what the log says. Thanks for the response, it definitely helped guide me in the right direction.
